spec/compiler/crystal/tools/doc/project_info_spec.cr

```diff
@@ -0,0 +1,128 @@
+require "../../../../spec_helper"
+require "../../../../support/tempfile"
+
+private alias ProjectInfo = Crystal::Doc::ProjectInfo
+
+describe Crystal::Doc::ProjectInfo do
+  around_each do |example|
+    with_tempfile("docs-project") do |tempdir|
+      Dir.mkdir tempdir
+      Dir.cd(tempdir) do
+        example.run
+      end
+    end
+  end
+
+  describe ".new_with_defaults" do
+    it "empty folder" do
+      ProjectInfo.new_with_defaults(nil, nil) { |name, version| "missing:#{name}:#{version}" }.should eq "missing::"
+      ProjectInfo.new_with_defaults("foo", "1.0") { |name, version| "missing:#{name}:#{version}" }.should eq ProjectInfo.new("foo", "1.0")
+    end
+
+    it "shard.yml with name and version" do
+      File.write("shard.yml", "name: foo\nversion: 1.0")
+
+      ProjectInfo.new_with_defaults(nil, nil) { |name, version| "missing:#{name}:#{version}" }.should eq ProjectInfo.new("foo", "1.0")
+      ProjectInfo.new_with_defaults("bar", "2.0") { |name, version| "missing:#{name}:#{version}" }.should eq ProjectInfo.new("bar", "2.0")
+      ProjectInfo.new_with_defaults(nil, "2.0") { |name, version| "missing:#{name}:#{version}" }.should eq ProjectInfo.new("foo", "2.0")
+    end
+
+    it "shard.yml and git tagged version" do
+      File.write("shard.yml", "name: foo\nversion: 1.0")
+      `git init`
+      `git add shard.yml`
+      `git commit -m 'Initial commit' --no-gpg-sign`
+      `git tag v3.0`
+
+      ProjectInfo.new_with_defaults(nil, nil) { |name, version| "missing:#{name}:#{version}" }.should eq ProjectInfo.new("foo", "3.0")
+      ProjectInfo.new_with_defaults("bar", "2.0") { |name, version| "missing:#{name}:#{version}" }.should eq ProjectInfo.new("bar", "2.0")
+    end
+
+    it "shard.yml and git tagged version dirty" do
+      File.write("shard.yml", "name: foo\nversion: 1.0")
+      File.write("foo.txt", "bar")
+      `git init`
+      `git add shard.yml`
+      `git commit -m 'Initial commit' --no-gpg-sign`
+      `git tag v3.0`
+
+      ProjectInfo.new_with_defaults(nil, nil) { |name, version| "missing:#{name}:#{version}" }.should eq ProjectInfo.new("foo", "1.0")
```

Comment:
This case needs some refinement. It feels wrong to simply fall back to the shard.yml version if the git repo is dirty. This is likely incorrect and a source of confusion. To make a dirty repo more explicit, I'd suggest to append -dev to the version (or some other indicator if there are better suggestions). In that specific case, I would still use the tag from HEAD (and append -dev) because git tags should have priority over shard.yml property.
Shard.yml version directly should only be used when we're not in a git repo or HEAD is not tagged and workdir is clean.

Reply:
> when we're not in a git repo or HEAD is not tagged and workdir is clean.

On second though, that shouldn't use the version from shard.yml either. It usually points to the latest release and doesn't represent the current state. Probably best to use the branch name then.